zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/controller/DocumentController.java

```diff
@@ -0,0 +1,115 @@
+ */
+package io.camunda.zeebe.gateway.rest.controller;
+
+import io.camunda.service.DocumentServices;
+import io.camunda.service.DocumentServices.DocumentReferenceResponse;
+import io.camunda.zeebe.gateway.protocol.rest.DocumentLinkRequest;
+import io.camunda.zeebe.gateway.protocol.rest.DocumentMetadata;
+import io.camunda.zeebe.gateway.protocol.rest.DocumentReference;
+import io.camunda.zeebe.gateway.rest.RequestMapper;
+import io.camunda.zeebe.gateway.rest.RestErrorMapper;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.CompletableFuture;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RequestPart;
+import org.springframework.web.multipart.MultipartFile;
+import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBody;
+
+@CamundaRestController
+@RequestMapping("/v2/documents")
+public class DocumentController {
+
+  private final DocumentServices documentServices;
+
+  @Autowired
+  public DocumentController(final DocumentServices documentServices) {
+    this.documentServices = documentServices;
+  }
+
+  @PostMapping
+  public CompletableFuture<ResponseEntity<DocumentReference>> createDocument(
+      @RequestParam(required = false) String documentId,
+      @RequestParam(required = false) String storeId,
+      @RequestPart("file") MultipartFile file,
+      @RequestPart(value = "metadata", required = false) DocumentMetadata metadata) {
+
+    final var enrichedMetadata = enrichMetadata(metadata, file);
+    return RequestMapper.toDocumentCreateRequest(documentId, storeId, enrichedMetadata)
+        .fold(
+            RestErrorMapper::mapProblemToCompletedResponse,
+            req -> createDocument(req, file, enrichedMetadata));
+  }
+
+  private CompletableFuture<ResponseEntity<DocumentReference>> createDocument(
+      final DocumentServices.DocumentCreateRequest request, final MultipartFile file, final DocumentMetadata metadata) {
+    final InputStream inputStream;
+    try {
+      inputStream = file.getInputStream();
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to read document content", e);
+    }
+    return documentServices.createDocument(request, inputStream)
+        .thenApply(resp -> toDocumentReference(resp, metadata))
+        .thenApply(reference -> new ResponseEntity<>(reference, HttpStatus.CREATED));
+  }
```

Comment:
Yep, I agree, that part could move into the services 👍

Reply:
I couldn't reuse the OpenAPI models directly or return them directly from the service since the service layer doesn't depend on the generated classes. But I created internal lookalike models and now the service consumes all information in one request and also produces the complete set of data required to convert it into an external reference. I agree that it's a good idea, and this will potentially allow services to add more metadata during document creation and return it to the client

Reply:
This is the way! 👨‍🚀